### DISCUSSION TITLE
fset return found value

### DISCUSSION THREAD

**Discussion by PowiHugh** [pymeasure/pymeasure#641] [Q&A]
Hello,

I'm setting up a few instruments (will send up the new instrument scripts later if allowed, I promise).  As part of the development I'm learning both pymeasure and python simultaneously, so I'm iterating through a lot of revisions of data types and function structures.

One of the things I'd like to implement is the ability to fset one function based on a known state of the equipment from prior fset inputs, e.g.:
1) Tell an arbitrary waveform generator "BSWV WVTP,SINE" translated: Initiate basic waveform, wavetype SINE.
2) At the end of the fset that established the AWG was going to use a SINE wave, store the state of the AWG waveform ("SINE") somewhere.
3) (The useful part) when executing the command "BSWV FRQ,%d", translated: Set the basic waveform frequency to %dvalue, have the value lambda through a function which, based on the waveform state, defines the upper limit of the frequency.  A BK Precision AWG (just off the top of my head, definitely not the specific device for which I'm developing right now) has three different upper limits depending on waveform type: SINE, 10-60Mhz (depending on model number); Arb, 12.5Mhz; Triangle, 500Khz.

Setting up all instruments to be supported by the same code with simple saved state and model information would be great.  This requires, however, that the successful execution of fset functions which change the state of the instrument return the value of the updated state variable.  Is there a reason that fset, in pymeasure right now, doesn't return anything?

Thanks in advance.

**Comment by LongnoseRob** [comment]
If the instrument is *talkative*, will say can be also be queried and reportes errors, then I would honestly not mess with trying to keep state of the instrument. 

I would then rather use some logic during the frequency set procedure, and check the mode property and assign the limits for the setting command based on the returned operational mode and a small LUT

Another possibility would be to not use such logic and just use the `check_set_errors` possibility in the instrument.setting [property](https://pymeasure.readthedocs.io/en/latest/api/instruments/instruments.html#pymeasure.instruments.Instrument.setting)

I am still working on a way of implmenting a base class for a series of instruments to support the same command set but different modell-number based specifications (like a family of power supplies sharing the same interface description but various output parameters)

**Reply by bilderbuchi** [reply]
> I would then rather use some logic during the frequency set procedure, and check the mode property and assign the limits for the setting command based on the returned operational mode and a small LUT

Yes, this! For the point 3 above,
* define a [dynamic](https://pymeasure.readthedocs.io/en/latest/dev/adding_instruments.html#dynamic-properties) `control` for the frequency with some limits (e.g. `strict_range`), 
* define a _method_ to set the waveform
* In that method, after setting the waveform, adjust the limits of your `frequency` property accordingly.

A minor point: `fset` has nothing to do with pymeasure, this is Python's way of designing the setter of a property. It does not make sense that this returns anything.

**Comment by msmttchr** [comment]
I think property's `fset` should return `None` by definition.
`pymeasure` offers also the, so called, dynamic properties, where you can change limits at class level (for limits that depends on instrument model) or at instance level (for limits that depends on instrument state).

Have a look at:
https://pymeasure.readthedocs.io/en/latest/dev/adding_instruments.html#dynamic-properties

Also instrument can be brought on a particular state by issuing commands via property or simple `write` or front panel, so I think it is difficult to track instrument state reliably.

Hope this helps

**Comment by PowiHugh** [comment]
I've read the replies so far and I recognize that non-pymeasure events (or even unanticipated pymeasure events) could change the state of the device such that the recorded state in the module becomes stale.

I like the idea that as part of setting certain functions, a get was required first.  I haven't seen (or possibly recognized, still learning a lot about classes and methods) a way to 'require a function as the first thing that happens before a setting function is executed' within the exist framework.

Am I missing something?

I should follow up here.  I read the verbiage on "set_process" more carefully, and I understand that I can execute a lambda expression prior to mapping valid value sets.  What that doesn't do is execute an instrument.measure function and set or return a state variable, from which I would then execute a lookup to determine a valid value boundary (maybe I can to that as a lamda function, but eugh.)  Reading the instrument.py code again I believe set_process and get_process options can **only** be lambda functions, correct?

**Reply by bilderbuchi** [reply]
set_process and get_process are meant for modifying the messages sent to and fro, not to trigger additional communication.
If you need to pack more logic than the current framework (properties/`control`) foresees, a method is typically appropriate. See my reply above, and also Michele's reply to basically the same effect.

**Reply by PowiHugh** [reply]
Yep, I think I finally understand the...logic? behind defining variable names as concatenations of other pieces of a class as a way of dynamically setting them after a function.  I don't have a specific instrument to try the following code just yet, but here is what I think is supposed to be the structure of this approach (according to what I've read):
Also note that I'm cribbing some of the code from the agilent 34450 DMM for this example.
----------------------------------------------------------------------------------------------------------------------------------
class dmm_generic(Instrument):
    BOOLS = {True: 1, False: 0}

    MODES = {'current': 'CURR', 'ac current': 'CURR:AC',
             'voltage': 'VOLT', 'ac voltage': 'VOLT:AC',
             'resistance': 'RES', '4w resistance': 'FRES',
             'current frequency': 'FREQ:ACI', 'voltage frequency': 'FREQ:ACV',
             'continuity': 'CONT',
             'diode': 'DIOD',
             'temperature': 'TEMP',
             'capacitance': 'CAP'}
    RANGES = {'current': [],
              'voltage': [100E-3, 1, 10, 100, 1000, "MAX", "MIN", "DEF"],
              'resistance': [100, 1E3, 10E3, 100E3, 1E6, 10E6, 100E6, "MAX", "MIN", "DEF"],
               '4w resistance': [100, 1E3, 10E3, 100E3, 1E6, 10E6, 100E6, "MAX", "MIN", "DEF"],
               'voltage frequency': [100E-3, 1, 10, 100, 750, "MAX", "MIN", "DEF"],
               'capacitance': [1E-9, 10E-9, 100E-9, 1E-6, 10E-6, 100E-6, 1E-3, 10E-3, "MAX", "MIN", "DEF"]}
    def __init__(self, adapter, **kwargs):
        super().__init__(
            adapter, 
            "Generic DMM framework", 
            **kwargs
        )

    @property
    def mode(self):
        get_command = ":configure?"
        vals = self._conf_parser(self.values(get_command))
        # Return only the mode parameter
        inv_modes = {v: k for k, v in self.MODES.items()}
        mode = inv_modes[vals[0]]
        return mode

    @mode.setter
    def mode(self, value):
        """ A string parameter that sets the measurement mode of the multimeter. Can be "current",
        "ac current", "voltage", "ac voltage", "resistance", "4w resistance", "current frequency",
        "voltage frequency", "continuity", "diode", "temperature", or "capacitance"."""
        if value in self.MODES:
            if value not in ['current frequency', 'voltage frequency']:
                self.write(':configure:' + self.MODES[value])
            else:
                if value == 'current frequency':
                    self.mode = 'ac current'
                else:
                    self.mode = 'ac voltage'
                self.write(":configure:freq")
        else:
            raise ValueError(f'Value {value} is not a supported mode for this device.')

...more stuff, not relevant...

    meas_range = Instrument.control(
        "RANG?",
        "RANG %s",
        """ Set or verify the present range of the mode to be measured. """,
        validator=strict_discrete_set,
        values=[1E-6, 10E-6, 100E-6, 1E-3, 10E-3, 100E-3, 1, 3, "MIN", "DEF", "MAX"],
        dynamic = True,
    )
    def set_range_details(self):
        self.mode
        meas_range_set_command = self.MODES[self.mode] + meas_range.set_command
        meas_range_get_command = self.MODES[self.mode] + meas_range.get_command
       meas_range_values = self.RANGES[self.mode]
----------------------------------------------------------------------------------------------------------------------------------
So assuming I didn't make any logic errors up top defining my dictionaries, copying (shamelessly) the mode properties and methods, and defining meas_range and set_range_details, I can now:
> define a single control, meas_range
> then use the new dynamic properties meas_range_set_command, meas_range_get_command and meas_range_values to, using the above dictionaries
> prepend some mode-specific verbiage to the command strings and
> redefine the values definition.

I'm developing in VSCode, and this program is NOT happy with this syntax.  Maybe I need an additional plugin.

Is this really considered valid Python syntax and structure?

Thanks in advance,
-Hugh

**Reply by bilderbuchi** [reply]
> Yep, I think I finally understand the...logic? behind defining variable names as concatenations of other pieces of a class as a way of dynamically setting them after a function.

yes, this was the best reasonable way to do this.

```
def mode(self, value):
    ...
      self.mode = 'ac current'
```
You have a circle here, this setter calls itself, which will blow up.

Yes, the ideas are there. Actually, all you need to do is in your definition of the `mode` setter, to add
```
@mode.setter
def mode(self, value):
    ...all the rest
    # The following updates the contents of values (i.e. admissible options for the validator) of the meas_range control based on the requested mode
    self.meas_range_values = self.RANGES[value]  # note the "self.", that was missing in your implementation.

**Reply by PowiHugh** [reply]
Ok, I can see how mapping ranges according to a new call to self.mode will work.

I then attempted to update set_command and get_command strings to prepend the specific mode for e.g. range:
 >       self.meas_range_values = self.RANGES[mode]
 >       self.__meas_range_set_command = self.MODES[mode] + self.__meas_range_set_command
 >       self.meas_range_get_command = self.MODES[mode] + self.meas_range_get_command

 but I run into this bit of code in instrument.py:

    def __getattribute__(self, name):
>        """ Prevent read access to variables with special names used to
>       support dynamic property behaviour """
>        if name in ('_special_names', '__dict__'):
>            return super().__getattribute__(name)
>        if hasattr(self, '_special_names'):
>            if name in self._special_names:
>                raise AttributeError(
>                    f"{name} is a reserved variable name and it cannot be read")
>        return super().__getattribute__(name)

I assume that there is a very good reason to reserve the dynamic property names as write-only, or internal-use-only.  This does make it basically impossible to define a "do_task_generically" property and then use context switching to dynamically (!!!) make it a "do_task_specifically" property.

What are the dangers of allowing these "special names" to be readable?

Thanks for all of this, I think I'm either getting to a solution or a good reason why this schema won't work.

**Reply by bilderbuchi** [reply]
```
  self.__meas_range_set_command = self.MODES[mode] + self.__meas_range_set_command
  self.meas_range_get_command = self.MODES[mode] + self.meas_range_get_command
```
I'm not sure if this does what you think it does. This will prepend the mode string _every time_ you hit this code path. So if your `self.meas_range_get_command` is first `bla`, this will become `VOLTbla`, `VOLTVOLTbla`, `VOLTVOLTVOLTbla`, etc.

As the commands for meas_range apparently _need_ the mode prefix, your control as defined above (`"RANG?"`/`"RANG %s"`) seems invalid. In your snippet, you could do
```
  self.meas_range_set_command = self.MODES[mode] + "RANG %s"
  self.meas_range_get_command = self.MODES[mode] + "RANG?"
```
which incidentally avoids reading the special attributes. If you want to avoid repetition, you could define module-level constants for those "trunks" and use them both in your `set_range_details` and in the definition of `meas_range`.

The reason for disallowing reads was, iirc, a much simpler logic for attribute handling, avoiding a circle somewhere, and limiting the dynamic features at first to avoid the worst footguns.
